lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -472,7 +472,12 @@ def _on_opensearch_data_storage_detaching(self, _: StorageDetachingEvent):  # no
                 "Removing units during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
             )
         # acquire lock to ensure only 1 unit removed at a time
-        if not self.node_lock.acquired:
+        # Closes canonical/opensearch-operator#378
+        if (
+            self.app.planned_units() > 1
+            and (self.opensearch.is_node_up() or self.alt_hosts)
+            and not self.node_lock.acquired
```

Comment:
What of the case where the nodes are down from a full cluster crash + scale-down for ex? why test on:
```
self.app.planned_units() > 1
and (self.opensearch.is_node_up() or self.alt_hosts)
```

Why not test on the number of units instead of the planned number?

Reply:
@Mehdi-Bendriss we can test on the node count, for the local peers of this app. The main issue here was the `self.node_lock.acquired` relying on the peer relation when we reach `node_count == 1`; and that peer relation will not exist.

Reply:
@Mehdi-Bendriss check [this comment](https://github.com/canonical/opensearch-operator/pull/361/files/e3588a23397c17f357a1d5ec3254958bca366fa1#r1711519631). What we want to know here is the following:
1) Are we on a `remove-application` case? If yes, then `self.app.planned_units == 0`, likewise for the last unit going away
2) Are we the last node to go? If yes, then there is no relevance in grabbing the lock anyways -> I was checking it with `self.app.planned_units > 1 `, which means we have at least 2 nodes "planned" 
3) Are we going to use the node_lock or the peer_lock, I am trying to avoid the later, so I have: `(self.opensearch.is_node_up() or self.alt_hosts)`.

> What of the case where the nodes are down from a full cluster crash + scale-down for ex?

In this case, `self.app.planned_units >= 1` will pass, then `(self.opensearch.is_node_up() or self.alt_hosts)` will fail, as no node is currently reachable. The lock won't be tried to be picked, the exception below will be skipped and the unit will go away simply, as planned.

Reply:
1. why won't the peer relation exist when `node_count == 1`? on_storage_detaching runs before any other teardown event https://github.com/canonical/opensearch-operator/pull/361#discussion_r1711396636
2. `planned_units == 1` can be even when there are currently 5 units and we just initiated a scale-down to 1, thus inaccurate

Reply:
@Mehdi-Bendriss I cannot answer the `why` question, as the peer relation object comes from ops and juju. One of these teams decided to not return that anymore post-`remove-application`. I can assure you that, in a `remove-application` and you're the last unit, then it will not exist.

```
2024-08-09 13:50:33,992 DEBUG    ops 2.15.0 up and running.
2024-08-09 13:50:34,065 DEBUG    Emitting Juju event opensearch_data_storage_detaching.
> /var/lib/juju/agents/unit-opensearch-0/charm/lib/charms/opensearch/v0/opensearch_base_charm.py(479)_on_opensearch_data_storage_detaching()
-> if not self.node_lock.acquired:
(Pdb) p self.model.get_relation(PeerRelationName)
None
```

Not sure what changed before and now. We did update ops to 2.15 and juju to 3.4.5. I am using juju 3.4.5, status: https://pastebin.ubuntu.com/p/YgVSbqWsGW/

Reply:
> planned_units == 1 can be even when there are currently 5 units and we just initiated a scale-down to 1, thus inaccurate

Sure, but I am protecting myself against the `remove-application`, hence I need the `planned_units`, not the current unit count.

Reply:
Dug into this issue and reported a bug to juju: https://bugs.launchpad.net/juju/+bug/2076599